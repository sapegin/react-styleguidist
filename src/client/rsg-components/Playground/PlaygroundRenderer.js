import React from 'react';
import PropTypes from 'prop-types';
import cx from 'clsx';
import Styled from 'rsg-components/Styled';

export const styles = ({ space, color, borderRadius }) => ({
	root: {
		marginBottom: space[4],
	},
	preview: {
		padding: space[2],
		border: [[1, color.border, 'solid']],
		borderRadius,
		// the next 2 lines are required to contain floated components
		width: '100%',
		display: 'inline-block',
	},
	controls: {
		display: 'flex',
		alignItems: 'center',
		marginBottom: space[1],
	},
	toolbar: {
		marginLeft: 'auto',
	},
	tab: {}, // expose className to allow using it in 'styles' settings
	padded: {
		// add padding between each example element rendered
		'& > *': {
			isolate: false,
			marginLeft: -space[1],
			marginRight: -space[1],
			'& > *': {
				isolate: false,
				marginRight: space[1],
				marginLeft: space[1],
			},
		},
	},
});

export function PlaygroundRenderer({
	classes,
	exampleIndex,
	name,
	padded,
	preview,
	previewProps,
	tabButtons,
	tabBody,
	toolbar,
}) {
	const { className, ...props } = previewProps;
	const previewClasses = cx(classes.preview, className, { [classes.padded]: padded });
	return (
		<div className={classes.root} data-testid={`${name}-example-${exampleIndex}`}>
			<div className={previewClasses} {...props} data-preview={name} data-testid="preview-wrapper">
				{preview}
			</div>
			<div className={classes.controls}>
				<div className={classes.tabs}>{tabButtons}</div>
				<div className={classes.toolbar}>{toolbar}</div>
			</div>
			<div className={classes.tab}>{tabBody}</div>
		</div>
	);
}

PlaygroundRenderer.propTypes = {
	classes: PropTypes.object.isRequired,
	exampleIndex: PropTypes.number.isRequired,
	name: PropTypes.string.isRequired,
	padded: PropTypes.bool.isRequired,
	preview: PropTypes.node.isRequired,
	previewProps: PropTypes.object.isRequired,
	tabButtons: PropTypes.node.isRequired,
	tabBody: PropTypes.node.isRequired,
	toolbar: PropTypes.node.isRequired,
};

export default Styled(styles)(PlaygroundRenderer);
