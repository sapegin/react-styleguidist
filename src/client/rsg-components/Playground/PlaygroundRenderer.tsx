import React from 'react';
import PropTypes from 'prop-types';
import cx from 'clsx';
import Styled, { JssInjectedProps } from 'rsg-components/Styled';
import * as Rsg from '../../../typings';

export const styles = ({ space, color, borderRadius }: Rsg.Theme) => ({
	root: {
		marginBottom: space[4],
	},
	preview: {
		padding: space[2],
		border: [[1, color.border, 'solid']],
		borderRadius,
		// the next 2 lines are required to contain floated components
		width: '100%',
		display: 'inline-block',
	},
	controls: {
		display: 'flex',
		alignItems: 'center',
		marginBottom: space[1],
	},
	toolbar: {
		marginLeft: 'auto',
	},
	tab: {}, // expose className to allow using it in 'styles' settings
	padded: {
		// add padding between each example element rendered
		'& > *': {
			isolate: false,
			marginLeft: -space[1],
			marginRight: -space[1],
			'& > *': {
				isolate: false,
				marginRight: space[1],
				marginLeft: space[1],
			},
		},
	},
});

interface PlaygroundRendererProps extends JssInjectedProps {
	exampleIndex: number | string;
	componentName: string;
	padded: boolean;
	preview: React.ReactNode;
	previewClassName?: string;
	tabButtons: React.ReactNode;
	tabBody: React.ReactNode;
	toolbar: React.ReactNode;
}

export const PlaygroundRenderer: React.FunctionComponent<PlaygroundRendererProps> = ({
	classes,
	exampleIndex,
	componentName,
	padded,
	preview,
	previewClassName,
	tabButtons,
	tabBody,
	toolbar,
}) => {
	const previewClasses = cx(classes.preview, previewClassName, { [classes.padded]: padded });
	return (
		<div className={classes.root} data-testid={`${componentName}-example-${exampleIndex}`}>
			<div className={previewClasses} data-preview={componentName} data-testid="preview-wrapper">
				{preview}
			</div>
			<div className={classes.controls}>
				<div className={classes.tabs}>{tabButtons}</div>
				<div className={classes.toolbar}>{toolbar}</div>
			</div>
			<div className={classes.tab}>{tabBody}</div>
		</div>
	);
};

PlaygroundRenderer.propTypes = {
	classes: PropTypes.objectOf(PropTypes.string.isRequired).isRequired,
	exampleIndex: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
	componentName: PropTypes.string.isRequired,
	padded: PropTypes.bool.isRequired,
	preview: PropTypes.node.isRequired,
	previewClassName: PropTypes.string,
	tabButtons: PropTypes.node.isRequired,
	tabBody: PropTypes.node.isRequired,
	toolbar: PropTypes.node.isRequired,
};

export default Styled<PlaygroundRendererProps>(styles)(PlaygroundRenderer);
