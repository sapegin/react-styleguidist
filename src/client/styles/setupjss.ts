import { create } from 'jss';
import global from 'jss-global';
import isolate from 'jss-isolate';
import nested from 'jss-nested';
import camelCase from 'jss-camel-case';
import defaultUnit from 'jss-default-unit';
import compose from 'jss-compose';
import nonInheritedProps from './nonInheritedProps';

const createGenerateId = () => {
	let counter = 0;
	return (rule:{key:string}) => `rsg--${rule.key}-${counter++}`;
};

const jss = create({
	createGenerateId,
	plugins: [
		global(),
		isolate({
			reset: {
				// Reset all inherited and non-inherited properties
				...nonInheritedProps,

				// “Global” styles for all components
				boxSizing: 'border-box',

				// Allow inheritance because it may be set on body and should be available for user components
				color: 'inherit',
				font: 'inherit',
				fontFamily: 'inherit',
				fontSize: 'inherit',
				fontWeight: 'inherit',
				lineHeight: 'inherit',
			},
		}),
		nested(),
		camelCase(),
		defaultUnit(),
		compose(),
	],
});

export default jss;
