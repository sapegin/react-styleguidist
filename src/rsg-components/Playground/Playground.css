.root {
	composes: border from "../../styles/colors.css";
	margin-bottom: 30px;
	border-radius: 3px;
}

.editor {
}

.preview {
	margin-bottom: 3px;
	padding: 15px;
}

.codeToggle {
	composes: font from "../../styles/common.css";
	composes: border from "../../styles/colors.css";
	composes: link from "../../styles/colors.css";
	cursor: pointer;
	display: inline-block;
	float: right;
	margin-bottom: -30px;
	margin-right: -1px;
	border-bottom-left-radius: 4px;
	border-bottom-right-radius: 4px;
	border-top: 0;
	padding: 4px 8px;
	font-size: small;
}

.showCode {
	composes: codeToggle;
	background: #fafafa;
}

.hideCode {
	composes: codeToggle;
	background: #f5f5f5;
}
