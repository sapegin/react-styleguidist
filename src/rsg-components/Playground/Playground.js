import React, { Component, PropTypes } from 'react';
import Editor from 'rsg-components/Editor';
import Preview from 'rsg-components/Preview';

import s from './Playground.css';

export default class Playground extends Component {
	static propTypes = {
		highlightTheme: PropTypes.string.isRequired,
		code: PropTypes.string.isRequired,
		evalInContext: PropTypes.func.isRequired,
		showCode: PropTypes.bool,
	};

	constructor(props) {
		super();
		this.state = {
			code: props.code,
			showCode: false,
		};
	}

	componentWillReceiveProps(nextProps) {
		let { code } = nextProps;
		if (code) {
			this.setState({
				code,
			});
		}
	}

	handleChange(newCode) {
		this.setState({
			code: newCode,
		});
	}

	handleCodeToggle() {
		this.setState({
			showCode: !this.state.showCode,
		});
	}

	render() {
		let { code } = this.state;
		let { highlightTheme } = this.props;

		return (
			<div className={s.root}>
				<div className={s.preview}>
					<Preview code={code} evalInContext={this.props.evalInContext} />
				</div>
				{this.state.showCode ? (
					<div className={s.editor}>
						<Editor code={code} highlightTheme={highlightTheme} onChange={code => this.handleChange(code)} />
						<div className={s.hideCode} onClick={() => this.handleCodeToggle()}>
							hide code
						</div>
					</div>
				) : (
					<div className={s.showCode} onClick={() => this.handleCodeToggle()}>
						show code
					</div>
				)}
			</div>
		);
	}
}
