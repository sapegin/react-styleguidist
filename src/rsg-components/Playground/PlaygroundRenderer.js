import React, { PropTypes } from 'react';
import Editor from 'rsg-components/Editor';
import Link from 'rsg-components/Link';
import Preview from 'rsg-components/Preview';
import Styled from 'rsg-components/Styled';

const styles = ({ font, link, linkHover, border, baseBackground, codeBackground }) => ({
	root: {
		position: 'relative',
		marginBottom: 30,
		border: [[1, border, 'solid']],
		borderRadius: '3px 3px 0 3px',
		'&:hover $isolatedLink': {
			isolate: false,
			opacity: 1,
		},
	},
	preview: {
		marginBottom: 3,
		padding: 15,
	},
	codeToggle: {
		position: 'absolute',
		right: -1,
		margin: 0,
		padding: [[6, 8]],
		fontFamily: font,
		fontSize: 14,
		lineHeight: 1,
		color: link,
		border: [[1, border, 'solid']],
		borderTop: 0,
		borderBottomLeftRadius: 3,
		borderBottomRightRadius: 3,
		cursor: 'pointer',
		'&:hover, &:active': {
			isolate: false,
			color: linkHover,
		},
	},
	showCode: {
		composes: '$codeToggle',
		backgroundColor: baseBackground,
	},
	hideCode: {
		composes: '$codeToggle',
		backgroundColor: codeBackground,
	},
	isolatedLink: {
		position: 'absolute',
		top: 0,
		right: 0,
		padding: [[6, 8]],
		fontFamily: font,
		fontSize: 14,
		opacity: 0,
		transition: 'opacity ease-in-out .15s .2s',
	},
});

export function PlaygroundRenderer({
	classes,
	code,
	showCode,
	name,
	index,
	isolatedExample,
	evalInContext,
	onChange,
	onCodeToggle,
}) {
	return (
		<div className={classes.root}>
			<div className={classes.preview} data-preview={name ? name : ''}>
				<div className={classes.isolatedLink}>
					{name && (
						isolatedExample ? (
							<Link href={'#!/' + name}>⇽ Exit Isolation</Link>
						) : (
							<Link href={'#!/' + name + '/' + index}>Open isolated ⇢</Link>
						)
					)}
				</div>
				<Preview code={code} evalInContext={evalInContext} />
			</div>
			{showCode ? (
				<div>
					<Editor code={code} onChange={onChange} />
					<button type="button" className={classes.hideCode} onClick={onCodeToggle}>
						Hide code
					</button>
				</div>
			) : (
				<button type="button" className={classes.showCode} onClick={onCodeToggle}>
					Show code
				</button>
			)}
		</div>
	);
}

PlaygroundRenderer.propTypes = {
	classes: PropTypes.object.isRequired,
	code: PropTypes.string.isRequired,
	showCode: PropTypes.bool.isRequired,
	index: PropTypes.number.isRequired,
	evalInContext: PropTypes.func.isRequired,
	onChange: PropTypes.func.isRequired,
	onCodeToggle: PropTypes.func.isRequired,
	name: PropTypes.string,
	isolatedExample: PropTypes.bool,
};

export default Styled(styles)(PlaygroundRenderer);
