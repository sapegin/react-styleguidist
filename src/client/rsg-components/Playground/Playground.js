import React, { Component } from 'react';
import PropTypes from 'prop-types';
import debounce from 'lodash/debounce';
import Preview from 'rsg-components/Preview';
import Para from 'rsg-components/Para';
import Slot from 'rsg-components/Slot';
import PlaygroundRenderer from 'rsg-components/Playground/PlaygroundRenderer';
import Context from 'rsg-components/Context';
import { EXAMPLE_TAB_CODE_EDITOR } from '../slots';
import { DisplayModes, ExampleModes } from '../../consts';

class Playground extends Component {
	static propTypes = {
		code: PropTypes.string.isRequired,
		evalInContext: PropTypes.func.isRequired,
		index: PropTypes.number.isRequired,
		name: PropTypes.string.isRequired,
		exampleMode: PropTypes.string.isRequired,
		settings: PropTypes.object,
	};
	static defaultProps = {
		settings: {},
	};
	static contextType = Context;

	handleChange = debounce(code => {
		this.setState({
			code,
		});
	}, this.context.config.previewDelay);

	state = {
		code: this.props.code,
		prevCode: this.props.code,
		activeTab: this.getInitialActiveTab() ? EXAMPLE_TAB_CODE_EDITOR : undefined,
	};

	static getDerivedStateFromProps(nextProps, prevState) {
		const { code } = nextProps;
		if (prevState.prevCode !== code) {
			return {
				prevCode: code,
				code,
			};
		}
		return null;
	}

	componentWillUnmount() {
		// Clear pending changes
		this.handleChange.cancel();
	}

	getInitialActiveTab() {
		const expandCode = this.props.exampleMode === ExampleModes.expand;
		return this.props.settings.showcode !== undefined ? this.props.settings.showcode : expandCode;
	}

	handleTabChange = name => {
		this.setState(state => ({
			activeTab: state.activeTab !== name ? name : undefined,
		}));
	};

	render() {
		const { code, activeTab } = this.state;
		const { evalInContext, index, name, settings, exampleMode } = this.props;
		const { displayMode } = this.context;
		const isExampleHidden = exampleMode === ExampleModes.hide;
		const isEditorHidden = settings.noeditor || isExampleHidden;
		const preview = <Preview code={code} evalInContext={evalInContext} />;

		return isEditorHidden ? (
			<Para>{preview}</Para>
		) : (
			<PlaygroundRenderer
				name={name}
				exampleIndex={index}
				padded={!!settings.padded}
				preview={preview}
				previewProps={settings.props || {}}
				tabButtons={
					<Slot
						name="exampleTabButtons"
						active={activeTab}
						props={{ onClick: this.handleTabChange }}
					/>
				}
				tabBody={
					<Slot
						name="exampleTabs"
						active={activeTab}
						onlyActive
						// evalInContext passed through to support custom slots that eval code
						props={{ code, onChange: this.handleChange, evalInContext }}
					/>
				}
				toolbar={
					<Slot
						name="exampleToolbar"
						props={{ name, isolated: displayMode === DisplayModes.example, example: index }}
					/>
				}
			/>
		);
	}
}

export default Playground;
