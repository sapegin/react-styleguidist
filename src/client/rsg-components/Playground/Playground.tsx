import React, { Component } from 'react';
import PropTypes from 'prop-types';
import debounce from 'lodash/debounce';
import Preview from 'rsg-components/Preview';
import Para from 'rsg-components/Para';
import Slot from 'rsg-components/Slot';
import PlaygroundRenderer from 'rsg-components/Playground/PlaygroundRenderer';
import Context from 'rsg-components/Context';
import { EXAMPLE_TAB_CODE_EDITOR } from '../slots';
import { getIsolatedUrl } from '../../utils/getUrl';
import { ExampleModes } from '../../consts';
import * as Rsg from '../../../typings';

const getIsolatedButtonUrl = ({
	isolated,
	hashPath,
	exampleIndex,
}: {
	isolated: boolean;
	hashPath: string[];
	exampleIndex?: number | string;
}) => {
	if (isolated) {
		return getIsolatedUrl(hashPath);
	} else {
		return getIsolatedUrl(hashPath, exampleIndex);
	}
};

interface PlaygroundProps {
	componentName: string;
	componentHashPath: string[];
	exampleMode: string;
	code: string;
	documentScope: Record<string, unknown>;
	exampleScope: Record<string, unknown>;
	modifiers: Rsg.Modifiers;
}

interface PlaygroundState {
	code: string;
	prevCode: string;
	activeTab?: string;
}

class Playground extends Component<PlaygroundProps, PlaygroundState> {
	public static propTypes = {
		componentName: PropTypes.string.isRequired,
		componentHashPath: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
		exampleMode: PropTypes.string.isRequired,
		code: PropTypes.string.isRequired,
		documentScope: PropTypes.object.isRequired,
		exampleScope: PropTypes.object.isRequired,
		modifiers: PropTypes.object.isRequired,
	};

	public static contextType = Context;

	private handleChange = debounce((code) => {
		this.setState({
			code,
		});
	}, this.context.config.previewDelay);

	public state: PlaygroundState = {
		code: this.props.code,
		prevCode: this.props.code,
		activeTab: this.getInitialActiveTab() ? EXAMPLE_TAB_CODE_EDITOR : undefined,
	};

	public static getDerivedStateFromProps(nextProps: PlaygroundProps, prevState: PlaygroundState) {
		const { code } = nextProps;
		if (prevState.prevCode !== code) {
			return {
				prevCode: code,
				code,
			};
		}
		return null;
	}

	public componentWillUnmount() {
		// Clear pending changes
		this.handleChange.cancel();
	}

	private getInitialActiveTab(): boolean {
		const { showcode } = this.props.modifiers;
		const expandCode = this.props.exampleMode === ExampleModes.expand;
		return showcode !== undefined ? showcode : expandCode;
	}

	private handleTabChange = (name: string) => {
		this.setState((state) => ({
			activeTab: state.activeTab !== name ? name : undefined,
		}));
	};

	public render() {
		const { code, activeTab } = this.state;
		const {
			documentScope,
			exampleScope,
			componentName,
			componentHashPath,
			modifiers,
			exampleMode,
		} = this.props;
		const { isolated, exampleIndex } = this.context;
		const isExampleHidden = exampleMode === ExampleModes.hide;
		const isEditorHidden = modifiers.noeditor || isExampleHidden;
		const isolatedButton = isolated && exampleIndex !== undefined;

		const preview = (
			<Preview code={code} documentScope={documentScope} exampleScope={exampleScope} />
		);

		return isEditorHidden ? (
			<Para>{preview}</Para>
		) : (
			<PlaygroundRenderer
				componentName={componentName}
				exampleIndex={modifiers.index}
				padded={!!modifiers.padded}
				preview={preview}
				previewClassName={modifiers['preview-class']}
				tabButtons={
					<Slot
						name="exampleTabButtons"
						active={activeTab}
						props={{ onClick: this.handleTabChange }}
					/>
				}
				tabBody={
					<Slot
						name="exampleTabs"
						active={activeTab}
						onlyActive
						props={{ code, onChange: this.handleChange }}
					/>
				}
				toolbar={
					<Slot
						name="exampleToolbar"
						props={{
							name: componentName,
							href: getIsolatedButtonUrl({
								isolated: isolatedButton,
								hashPath: componentHashPath,
								exampleIndex: modifiers.index,
							}),
							exampleIndex: modifiers.index,
							isolated: isolatedButton,
						}}
					/>
				}
			/>
		);
	}
}

export default Playground;
