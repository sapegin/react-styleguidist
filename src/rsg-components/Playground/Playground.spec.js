import React from 'react';
import Playground from './Playground';
import '../../styles/setupjss';
import { PlaygroundRenderer } from './PlaygroundRenderer';

const code = '<button>OK</button>';
const newCode = '<button>Not OK</button>';
const options = {
	context: {
		config: {
			showCode: false,
			highlightTheme: 'base16-light',
		},
	},
};

it('should render component renderer', () => {
	const actual = shallow(
		<Playground code={code} evalInContext={a => () => a} name="name" index={0} />,
		options
	);

	expect(actual).toMatchSnapshot();
});

it('should update code', () => {
	const actual = shallow(
		<Playground code={code} evalInContext={a => () => a} name="name" index={0} />,
		options
	);

	expect(actual.prop('code')).toEqual(code);

	actual.instance().handleChange(newCode);

	expect(actual.prop('code')).toEqual(newCode);
});

it('should update code via props', () => {
	const actual = shallow(
		<Playground code={code} evalInContext={a => () => a} name="name" index={0} />,
		options
	);

	expect(actual.state('code')).toEqual(code);

	actual.setProps({
		code: newCode,
	});

	expect(actual.state('code')).toEqual(newCode);
});

it('should update code with debounce', done => {
	const actual = shallow(<Playground code={code} evalInContext={a => a} name="name" index={0} />, {
		context: {
			config: {
				...options.context.config,
				previewDelay: 1,
			},
		},
	});

	expect(actual.prop('code')).toEqual(code);

	actual.instance().handleChange(newCode);

	expect(actual.prop('code')).toEqual(code);
	setTimeout(() => {
		expect(actual.prop('code')).toEqual(newCode);
		done();
	}, 5);
});

it('should open a code editor', () => {
	const actual = mount(
		<Playground code={code} evalInContext={a => () => a} name="name" index={0} />,
		options
	);

	expect(actual.find('.ReactCodeMirror')).toHaveLength(0);

	actual.find('button').simulate('click');

	expect(actual.find('.ReactCodeMirror')).toHaveLength(1);
});

it('renderer should render preview', () => {
	const actual = shallow(
		<PlaygroundRenderer
			classes={{}}
			code={code}
			showCode={false}
			evalInContext={a => () => a}
			name="name"
			index={0}
			onChange={() => {}}
			onCodeToggle={() => {}}
		/>
	);

	expect(actual).toMatchSnapshot();
});
