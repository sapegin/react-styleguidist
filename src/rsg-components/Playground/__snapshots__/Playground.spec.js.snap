// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderer should render preview 1`] = `
<div>
  <div
    data-preview="name"
  >
    <div>
      <_class
        href="#!/name/0"
      >
        Open isolated â‡¢
      </_class>
    </div>
    <Preview
      code="<button>OK</button>"
      evalInContext={[Function]}
    />
  </div>
  <button
    onClick={[Function]}
    type="button"
  >
    Show code
  </button>
</div>
`;

exports[`should render component renderer 1`] = `
<_class
  code="<button>OK</button>"
  evalInContext={[Function]}
  index={0}
  isolatedExample={false}
  name="name"
  onChange={[Function]}
  onCodeToggle={[Function]}
  showCode={false}
/>
`;
